### PR TITLE
add report package

### PR DESCRIPTION
**Pull request by LinuxGit**
add report package: fetch images in parrallel form Grafana sever and render to a PDF file.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -10,7 +10,7 @@ PACKAGES := $$(go list ./... | grep -vE 'vendor')
 build: check test grafana_collector
 
 grafana_collector:
-	$(GO) build -ldflags '$(LDFLAGS)' -o bin/grafana_collector ./metrics/grafana_collector/cmd/grafana_collector/*.go
```
Comment (GregoryIan) [inline Makefile]: just use directory `cmd/grafana_collector/*.go`, remove parent directory `/metrics` and `/metrics/grafana_collector`

File: grafana_collector/report/report.go [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+
+package report
+
+import (
+	"fmt"
+	"github.com/ngaut/log"
+	"github.com/pborman/uuid"
+	"github.com/pingcap/tidb-inspect-tools/grafana_collector/grafana"
+	"github.com/signintech/gopdf"
+	"io"
```
Comment (GregoryIan) [inline grafana_collector/report/report.go]: separator built-in package and third-party package

File: grafana_collector/report/report.go [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+
+// Generate returns the report.pdf file. After reading this file it should be Closed()
+// After closing the file, call report.Clean() to delete the file
+func (rep *report) Generate() (pdf io.ReadCloser, err error) {
+	dash, err := rep.gClient.GetDashboard(rep.dashName)
+	if err != nil {
+		err = fmt.Errorf("error fetching dashboard %v: %v", rep.dashName, err)
+		return
+	}
+	err = rep.renderPNGsParallel(dash)
```
Comment (GregoryIan) [inline grafana_collector/report/report.go]: add blank line between  two block of logic code

File: grafana_collector/report/report.go [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+	if err != nil {
+		err = fmt.Errorf("error rendering PNGs in parralel for dash %+v: %v", dash, err)
+		return
+	}
+	pdf, err = rep.renderPDF(dash)
```
Comment (GregoryIan) [inline grafana_collector/report/report.go]: it's better to use error trace

Comment (GregoryIan) [reply grafana_collector/report/report.go]: use package `gitHub.com/juju/errors`

File: grafana_collector/report/report.go [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+}
+
+func (rep *report) renderPNG(p grafana.Panel) error {
+	body, err := rep.gClient.GetPanelPng(p, rep.dashName, rep.time)
+	if err != nil {
+		return fmt.Errorf("error getting panel %+v: %v", p, err)
+	}
+	defer body.Close()
+
+	err = os.MkdirAll(rep.imgDirPath(), 0777)
```
Comment (GregoryIan) [inline grafana_collector/report/report.go]: why not to call mkdirAll once in prepare phase of `Generate`?

File: grafana_collector/report/report.go [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+		return fmt.Errorf("error copying body to file:%v", err)
+	}
+	return nil
+}
+
+func (rep *report) renderPDF(dash grafana.Dashboard) (outputPDF *os.File, err error) {
+	pdf := gopdf.GoPdf{}
+	pdf.Start(gopdf.Config{PageSize: gopdf.Rect{W: 595.28, H: 841.89}})
+	pdf.AddPage()
+	err = pdf.AddTTFFont("opensans", "ttf/OpenSans-Regular.ttf")
```
Comment (GregoryIan) [inline grafana_collector/report/report.go]: make arguments about pdf generation configurable

File: grafana_collector/report/report.go [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+// New creates a new Report
+func New(g grafana.Client, dashName string, time grafana.TimeRange) Report {
+	return new(g, dashName, time)
+}
+
+func new(g grafana.Client, dashName string, time grafana.TimeRange) *report {
+	tmpDir := filepath.Join("tmp", uuid.New())
+	return &report{g, time, dashName, tmpDir}
+}
+
```
Comment (zhouqiang-cl) [inline grafana_collector/report/report.go]: Why New call new

File: grafana_collector/report/report.go [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+	_, err = io.Copy(file, body)
+	if err != nil {
+		return fmt.Errorf("error copying body to file:%v", err)
+	}
+	return nil
+}
+
+func (rep *report) renderPDF(dash grafana.Dashboard) (outputPDF *os.File, err error) {
+	pdf := gopdf.GoPdf{}
```
Comment (GregoryIan) [inline grafana_collector/report/report.go]: extract code related to config pdf  to a function and add some comments. some reviewers can review it carelessly

File: grafana_collector/report/report.go [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+
+	panels := make(chan grafana.Panel, len(dash.Panels))
+	for _, p := range dash.Panels {
+		panels <- p
+	}
+	close(panels)
+
+	var wg sync.WaitGroup
+	var count int
+	wg.Add(1)
```
Comment (GregoryIan) [inline grafana_collector/report/report.go]: non-concurrent, u don't need these synchronization and concurrency logic

File: grafana_collector/report/report.go [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+
+	//fetch images in parrallel form Grafana sever.
+	//limit concurrency using a worker pool to avoid overwhelming grafana
+	//for dashboards with many panels.
+	var wg sync.WaitGroup
+	workers := 5
+	wg.Add(workers)
+	errs := make(chan error, len(dash.Panels)) //routines can return errors on a channel
+	for i := 0; i < workers; i++ {
+		go func(panels <-chan grafana.Panel, errs chan<- error) {
```
Comment (zhouqiang-cl) [inline grafana_collector/report/report.go]: can extract this func out

File: grafana_collector/report/report.go [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+	close(panels)
+
+	var wg sync.WaitGroup
+	var count int
+	wg.Add(1)
+	errs := make(chan error, len(dash.Panels))
+	go func(panels <-chan grafana.Panel, errs chan<- error) {
+		defer wg.Done()
+		for p := range panels {
+			imgFileName := fmt.Sprintf("image%d.png", p.ID)
```
Comment (GregoryIan) [inline grafana_collector/report/report.go]: make 
```
imgFileName := fmt.Sprintf("image%d.png", p.ID)
imgFilePath := filepath.Join(rep.imgDirPath(), imgFileName)
```
a function

File: Makefile [hunk 12]
```diff
@@ -10,7 +10,7 @@ PACKAGES := $$(go list ./... | grep -vE 'vendor')
...
+	$(GO) build -ldflags '$(LDFLAGS)' -o bin/grafana_collector grafana_collector/cmd/grafana_collector/*.go
```
Comment (GregoryIan) [inline Makefile]: `$(GO) build -ldflags '$(LDFLAGS)' -o bin/grafana_collector cmd/grafana_collector/*.go`

File: grafana_collector/cmd/grafana_collector/main.go [hunk 13]
```diff
@@ -60,6 +64,16 @@ func main() {
 		return
 	}
 
+	err := report.NewConfig(*configFile)
```
Comment (GregoryIan) [inline grafana_collector/cmd/grafana_collector/main.go]: `report.SetConfig`

File: grafana_collector/cmd/grafana_collector/main.go [hunk 14]
```diff
@@ -60,6 +64,16 @@ func main() {
...
+	if err != nil {
+		log.Fatalf("parsing configure file error: %v", err)
+	}
+
+	if *fontDir == "" {
+		log.Fatalf("missing parameter: -font-dir")
+	}
+	report.NewFontDir(*fontDir)
```
Comment (GregoryIan) [inline grafana_collector/cmd/grafana_collector/main.go]: ditto

File: grafana_collector/report/report.go [hunk 15]
```diff
@@ -0,0 +1,252 @@
...
+	panels := make(chan grafana.Panel, len(dash.Panels))
+	for _, p := range dash.Panels {
+		panels <- p
+	}
+	close(panels)
+
+	//fetch images in parrallel form Grafana sever.
+	//limit concurrency using a worker pool to avoid overwhelming grafana
+	//for dashboards with many panels.
+	var wg sync.WaitGroup
```
Comment (GregoryIan) [inline grafana_collector/report/report.go]:
```
var (
    wg sync.WaitGroup
    workers = 5
    errs = make(chan error, len(dash.Panels))
)
```

File: grafana_collector/report/report.go [hunk 16]
```diff
@@ -0,0 +1,252 @@
...
+	workers := 5
+	wg.Add(workers)
+	errs := make(chan error, len(dash.Panels)) //routines can return errors on a channel
+	for i := 0; i < workers; i++ {
+		go func(panels <-chan grafana.Panel, errs chan<- error) {
+			defer wg.Done()
+			for p := range panels {
+				err := rep.renderPNG(p)
+				if err != nil {
+					log.Errorf("Error creating image for panel: %v", err)
```
Comment (GregoryIan) [inline grafana_collector/report/report.go]: `log.Errorf("Error creating image for panel %s: %v", xxx, err)`

File: grafana_collector/report/report.go [hunk 17]
```diff
@@ -0,0 +1,252 @@
...
+}
+
+func (rep *report) renderPNG(p grafana.Panel) error {
+	body, err := rep.gClient.GetPanelPng(p, rep.dashName, rep.time)
+	if err != nil {
+		return errors.Errorf("error getting panel %+v: %v", p, err)
+	}
+	defer body.Close()
+
+	imgFileName := fmt.Sprintf("image%d.png", p.ID)
```
Comment (GregoryIan) [inline grafana_collector/report/report.go]: use `imgDirPath `

File: grafana_collector/report/report.go [hunk 18]
```diff
@@ -0,0 +1,252 @@
...
+		}
+		if err != nil {
+			log.Errorf("Error rendering image to PDF: %v", err)
+		} else {
+			log.Infof("Rendering image to PDF: %s", imgPath)
+		}
+		count++
+	}
+
+	pdf.WritePdf(rep.pdfPath())
```
Comment (GregoryIan) [inline grafana_collector/report/report.go]: does it return error?

File: grafana_collector/report/report.go [hunk 19]
```diff
@@ -0,0 +1,252 @@
...
+	outputPDF, err = os.Open(rep.pdfPath())
+	return outputPDF, err
```
Comment (GregoryIan) [inline grafana_collector/report/report.go]: `return outputPDF, errors.Trace(err)`

**Pull request merged**